implementations/rust/ockam/ockam_command/src/run/parser.rs

```diff
@@ -0,0 +1,259 @@
+use crate::{CommandGlobalOpts, Result};
+use duct::Expression;
+use ockam_core::compat::collections::HashMap;
+use once_cell::sync::Lazy;
+use serde::Deserialize;
+use std::collections::{BTreeMap, VecDeque};
+use std::fmt::Debug;
+use std::path::Path;
+use tracing::debug;
+
+pub struct ConfigRunner {
+    commands_sorted: Vec<ParsedCommand>,
+    commands_index: BTreeMap<String, usize>,
+}
+
+#[derive(Clone)]
+pub struct ParsedCommand {
+    pub id: String,
+    pub depends_on: Option<String>,
+    pub cmd: Expression,
+}
+
+impl ConfigRunner {
+    fn new() -> Self {
+        Self {
+            commands_sorted: vec![],
+            commands_index: Default::default(),
+        }
+    }
+
+    pub fn go(opts: &CommandGlobalOpts, path: &Path) -> Result<()> {
+        let mut cr = Self::new();
+        cr.parse(opts, path)?;
+        cr.run()?;
+        Ok(())
+    }
+
+    fn parse(&mut self, opts: &CommandGlobalOpts, path: &Path) -> Result<()> {
+        let config = std::fs::read_to_string(path)?;
+        let config: Config = serde_yaml::from_str(&config)?;
+        let mut nodes = VecDeque::new();
+        for (name, node) in config.nodes {
+            nodes.push_back((name, node));
+        }
+        while let Some((name, node)) = nodes.pop_front() {
+            // If the node depends on another node, check if that node has been parsed.
+            if let Some(depends_on) = &node.depends_on {
+                // If the node hasn't been parsed, push it back to the queue.
+                if !self
+                    .commands_index
+                    .contains_key(&format!("node/{depends_on}"))
+                {
+                    nodes.push_back((name, node));
+                    continue;
+                }
+            }
+            // Else, proceed.
+            node.parse(opts, name, self)?;
+        }
+        Ok(())
+    }
+
+    fn run(self) -> Result<()> {
+        for c in self.commands_sorted.into_iter() {
+            debug!("Running command: {}", c.id);
+            // If a command fails it will show the appropriate error in its subshell.
+            // No need to return an error here.
+            if c.cmd.run().is_err() {
+                break;
+            }
+        }
+        Ok(())
+    }
+}
+
+/// The config structure will be a yml file with the following structure:
+/// ```yml
+/// nodes:
+///   telegraf:
+///     enrollment-token: $OCKAM_TELEGRAF_TOKEN
+///     tcp-inlets:
+///       telegraf:
+///         from: 127.0.0.1:8087
```

Comment:
Based on how often YAML parsing has injured itself in its confusion for string-like values that start with an integer, should we document these as:

```suggestion
///         from: '127.0.0.1:8087'
```

I've had some languages' parsers try to give me a float from IP addresses before.

Reply:
Good point, I'll do that, thanks for the heads up!